src/patchelf.cc

```diff
@@ -1033,31 +1022,33 @@ void ElfFile<ElfFileParamNames>::rewriteSectionsExecutable()
 
         firstPage -= neededPages * getPageSize();
         startOffset += neededPages * getPageSize();
-    } else {
-        Elf_Off rewrittenSectionsOffset = sizeof(Elf_Ehdr) + phdrs.size() * sizeof(Elf_Phdr);
-        for (auto& phdr : phdrs)
-            if (rdi(phdr.p_type) == PT_LOAD &&
-                rdi(phdr.p_offset) <= rewrittenSectionsOffset &&
-                rdi(phdr.p_offset) + rdi(phdr.p_filesz) > rewrittenSectionsOffset &&
-                rdi(phdr.p_filesz) < neededSpace)
-            {
-                wri(phdr.p_filesz, neededSpace);
-                wri(phdr.p_memsz, neededSpace);
-                break;
-            }
     }
 
+    Elf_Off currOffset = sizeof(Elf_Ehdr) + phdrs.size() * sizeof(Elf_Phdr);
 
-    /* Clear out the free space. */
-    Elf_Off curOff = sizeof(Elf_Ehdr) + phdrs.size() * sizeof(Elf_Phdr);
-    debug("clearing first %d bytes\n", startOffset - curOff);
-    memset(fileContents->data() + curOff, 0, startOffset - curOff);
+    /* Ensure PHDR is covered by a LOAD segment.
 
+       Because PHDR is supposed to have been covered by such section before, in
+       here we assume that we don't have to create any new section, but rather
+       extend the existing one. */
+    for (auto& phdr : phdrs)
```

Comment:
This logic used to be run only if `neededSpace <= startOffset`, under the assumption that otherwise (when `neededSpace > startOffset`), we call `shiftFile()`, which will automatically create appropriate `PT_LOAD` segment:

https://github.com/NixOS/patchelf/blob/7c2f768bf9601268a4e71c2ebe91e2011918a70f/src/patchelf.cc#L557

But _that_ was done under the assumption that `shiftFile()`'s `startOffset` will land inside a loadable segment that happens to contain `PHDR`, which is not necessarily true!  Some binaries have a dedicated small `LOAD` segment just for `PHDR`, making `shiftFile()` then pick the next available `LOAD` segment.

I believe this is the same case this merge request tried to cover (or at least both make `pcloud` a valid binary):

https://github.com/NixOS/patchelf/pull/264

This is now tested as a part of the `short-first-segment` test.

Reply:
Hmm, in a hindsight, now I'm thinking whether this approach can't cause some segments to overlap each other? (maybe that's why tests on a few other architectures fail, as I now see)